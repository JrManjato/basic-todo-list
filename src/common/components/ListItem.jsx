import React from 'react'

export default function ListItem({ id, text, isCheckable, setList, list }) {
    
    // Deletes an item based on the `item.id` key
    function deleteItem(id) {
        const newArray = list.filter((item) => item.id !== id);
        setList(newArray);
    }

    function updateScale(id) {
        // Get the current item
        const currentItem = list.filter((item) => item.id === id);
    
        // Create a new item with the updated scale
        const newItem = {
          id: currentItem[0].id,
          value: currentItem[0].value,
          scale: 'done'
        };
    
        deleteItem(id);
    
        // Update the list
        setList((oldList) => [...oldList, newItem]);
      }

    return (
        <li key={id}>
            {isCheckable &&
                <input
                    type="checkbox"
                    id={text}
                    onChange={() => updateScale(id)}
                />
            }
            <label htmlFor={text}>{text}</label>
        </li>
    )
}
