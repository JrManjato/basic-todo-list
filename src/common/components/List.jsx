import React from 'react';
import ListItem from './ListItem';

const List = ({ list, scale, isCheckable, setList }) => {
    return (
        <ul data-testid={`${scale}-list`}>
            {list.map((item) => item.scale === scale && (
                <ListItem
                    id={item.id}
                    text={item.value}
                    isCheckable={isCheckable}
                    setList={setList}
                    list={list}
                    scale={scale}
                    key={item.id}
                />
            ))}
        </ul>
    )
}

export default List;