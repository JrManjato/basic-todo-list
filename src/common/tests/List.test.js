import { create } from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import List from "../components/List";
import ListItem from "../components/ListItem";
import { MockList } from "./MockList";

describe("List Test", () => {

    test("should renders the component [snapshot]", () => {
        const todoList = create(<List list={[]} scale='todo' isCheckable={true} />);

        const json = todoList.toJSON();

        expect(json).toMatchSnapshot();
    });

    it("should return an empty list when rendered", () => {
        const { getByTestId } = render(<List list={[]} scale='todo' isCheckable={true} />);

        const todoList = getByTestId("todo-list");
        const listLength = todoList.children.length;

        expect(listLength).toBe(0);
    });

    it("should map the list according to task completion", () => {
        render(<List list={MockList} scale='todo' />);

        const listInTodo = screen.getByTestId("todo-list");
        const mockListInTodoLength = MockList.filter((item) => item.scale === 'todo').length;

        expect(listInTodo.children).toHaveLength(mockListInTodoLength);
    });

    it("should check that there are checkboxes when the status is todo", () => {
        render(<ListItem text isCheckable />);

        const listWitchCheckboxes = screen.getByTestId("checkbox");

        expect(listWitchCheckboxes).toBeTruthy();
    });
});