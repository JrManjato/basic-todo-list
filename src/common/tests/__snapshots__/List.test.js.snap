// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Test should renders the component [snapshot] 1`] = `
<ul
  data-testid="todo-list"
/>
`;
